Add cmac_pad testbench

minseongg opened the pull request:
Add testbench code for `cmac_pad` module. It includes two tests: `run_test_padding` and `run_test_tkeep`.

`run_test_padding` tests whether the size of the first packet is padded to 60.

`run_test_tkeep` tests whether `tdata` is masked well by `tkeep`.

alexforencich commented:
1. Please rebase and update on the current repo so that the CI tests run correctly (there is a bug in pytest-split I had to work around).  
2. What is the point of the tkeep test?  Generally you don't want to twiddle the tkeep signal manually with the AXI stream source/sink components; they're set up to remove all bytes with tkeep = 0 by default anyway, so I'm not sure what additional coverage that test case provides over the padding test.

minseongg commented:
Thank you for your quick review!

The reason I added the tkeep test is that I wanted to test below lines:

https://github.com/corundum/corundum/blob/b1c6bdbd88bd0c95d7f873ee1ee91fbb90a40f02/fpga/common/rtl/cmac_pad.v#L90-L96

But I think it is not needed if they're set up to remove all bytes with tkeep = 0 by default. I deleted the tkeep test.

alexforencich commented:
You know; I just did some thinking: a better test would be to use a short packet that gets padded, since the modules forces the first 60 tkeep bits to 1.  So in that case, the idea would be to generate, say, a 60 byte packet of non-zero bytes, override the default all-1s tkeep to make the packet significantly shorter than 60 bytes, then check that the resulting packet is 60 bytes long and that all of the pad bytes are zero.

alexforencich commented on fpga/common/tb/cmac_pad/test_cmac_pad.py:
```diff
@@ -0,0 +1,181 @@
+
+        # padded to 60 if the packet size is less than 60
+        if len(test_pkt) < 60:
+            padded_pkt = test_pkt + bytearray([0]*(60-len(test_pkt)))
```
How about you do this instead: get rid of the if statement and unconditionally pad test_pkt with `ljust` like so:

    padded_pkt = test_pkt.ljust(60, b'\x00')

minseongg commented:
I made some changes in the `run_test`.

Now `run_test` tests that padding works well when override the default all-1s tkeep with 60 bytes of tdata.

alexforencich commented:
Looks like you don't test anything larger than 60 bytes now, so you're not covering the case where the input is not padded.

alexforencich commented:
(I think you can simply bring back the previous test so you have two tests - one that sweeps over a bunch of input sizes, and the other that tests the zeroing)

minseongg commented:
Thanks. I brought back the previous test.

Now `run_test` sweeps over a bunch of input sizes, and `run_test_pad` tests the zeroing.
